[AKS] Changes to support MSI for IngressApplicationGateway addon

akshaysngupta opened the pull request:
This PR addresses the following:
1) Adds support for managed identity feature for AGIC addon. This means performing role assignment for AGIC addon identity once the cluster completes creation.
2) Modify help text for AGIC adding: "Used with IngressApplicationGateway addon"
3) Add tests for common scenarios when enabling AGIC addon with AKS.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).


### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

azuresdkci commented:
If this PR is for a new extension or change to an existing extension, use the following to try out the changes in this PR:
```
docker run -it microsoft/azure-cli:latest
export EXT=<NAME>
pip install --upgrade --target ~/.azure/cliextensions/$EXT "git+https://github.com/akshaysngupta/azure-cli-extensions.git@aksgupta/addon#subdirectory=src/$EXT&egg=$EXT"
```

yungezz commented:
hi @arrownj could you pls  help to review the PR? thanks.

arrownj commented on src/aks-preview/azext_aks_preview/custom.py:
```diff
@@ -1035,19 +1057,21 @@ def aks_create(cmd,     # pylint: disable=too-many-locals,too-many-statements,to
-                                    acr_name_or_id=attach_acr,
-                                    subscription_id=subscription_id)
+            if enable_managed_identity:
+                if attach_acr:
```
is there any special consideration here to split original one line condition check into two ?

arrownj replied:
Hi @akshaysngupta , do we need to merge this change?

akshaysngupta commented:
@arrownj I have updated the PR with tests.
Regarding your previous comment about "is there any special consideration here to split original one line condition check into two ?"

"monitoring" is an addon in AKS which when enabled needs the CLI to perform a role assignment once the cluster completes creation. The role assignment is done after the operation because the identity is created during the cluster creation. So these role assignments can't be done before the cluster creation.

We are making a similar change to IngressApplicationGateway addon. So, I have split the condition taking monitoring out.

```go
if monitoring
```
 to
```go
need_post_creation_role_assignment = monitoring or ingress_appgw_addon_enabled
if need_post_creation_role_assignment:
```

fengzhou-msft commented on src/aks-preview/azext_aks_preview/custom.py:
```diff
@@ -684,6 +684,60 @@ def _add_monitoring_role_assignment(result, cluster_resource_id, cmd):
+    service_principal_msi_id = None
+    # Check if service principal exists, if it does, assign permissions to service principal
+    # Else, provide permissions to MSI
+    if (
+            hasattr(result, 'service_principal_profile') and
+            hasattr(result.service_principal_profile, 'client_id') and
+            result.service_principal_profile.client_id != 'msi'
+    ):
+        service_principal_msi_id = result.service_principal_profile.client_id
+        is_service_principal = True
+    elif (
+            (hasattr(result, 'addon_profiles')) and
+            (CONST_INGRESS_APPGW_ADDON_NAME in result.addon_profiles) and
+            (hasattr(result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME], 'identity')) and
+            (hasattr(result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME].identity, 'object_id'))
+    ):
+        service_principal_msi_id = result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME].identity.object_id
+        is_service_principal = False
```
What if neither of condition is true?

And suggest to use `try` and `except` instead of multiple checks for the existence of attributes to follow [EAFP](https://docs.python.org/3/glossary.html#term-eafp)

arrownj replied:
Better to put this complicate condition check into a function such as `is_msi` ?

akshaysngupta replied:
added `if service_principal_msi_id is not None:`
